src/Strategy/Node/Npm/PackageLockV3.hs

```diff
@@ -0,0 +1,447 @@
+            PackageLockV3Root -> Nothing -- Don't report root project as dependency!
+        _ -> Nothing
+
+    allDependencyPackages :: Map PackagePath PackageLockV3Package
+    allDependencyPackages =
+      Map.filterWithKey
+        (\k _ -> isReportableDependencyModule k)
+        (packages pkgLockV3)
+
+    -- True if package path leads to dependency, Otherwise false.
+    isReportableDependencyModule :: PackagePath -> Bool
+    isReportableDependencyModule pkgPath = case pkgPath of
+      PackageLockV3PathKey _ _ -> True
+      _ -> False
+
+    -- Provides candidate path in top level tree
+    -- >> toTopLevelPackage pkgName = PackagePath "node_modules/" pkgName
+    toTopLevelPackage :: PackageName -> PackagePath
+    toTopLevelPackage = PackageLockV3PathKey defaultVendorDir
+
+    -- Provides candidate path in as if the package was vendored
+    -- >> toVendoredPackage "node_module/a" pkgName = PackagePath "node_modules/a/node_modules/" pkgName
+    toVendoredPackage :: Text -> PackageName -> PackagePath
+    toVendoredPackage prefix = PackageLockV3PathKey (prefix <> "/" <> defaultVendorDir)
+
+    -- Provides candidate path in current workspace root, if it contains a workspace root.
+    -- If prefix does not contain workspace root, it mimics @toVendoredPackage.
+    --
+    -- >> toVendoredPackage "myWorkspace/myPkg/node_modules/a" pkgName = PackagePath "myWorkspace/myPkg/node_modules/" pkgName
+    -- >> toVendoredPackage "myWorkspace/myPkg/node_modules/a/node_modules/b" pkgName = PackagePath "myWorkspace/myPkg/node_modules/" pkgName
+    -- >> toVendoredPackage "node_modules/a" pkgName = PackagePath "node_modules/a/node_modules/" pkgName
+    toWorkspaceLevelPackage :: Text -> PackageName -> PackagePath
+    toWorkspaceLevelPackage prefix = PackageLockV3PathKey (rootWorkspace <> "/" <> defaultVendorDir)
+      where
+        rootWorkspace :: Text
+        rootWorkspace = fromMaybe prefix (getRootWorkspace prefix)
+
+        getRootWorkspace :: Text -> Maybe Text
+        getRootWorkspace pkgPath = case Text.breakOn "/node_modules" pkgPath of
+          ("", "") -> Nothing
+          (firstModule, _) ->
+            if firstModule == pkgPath
+              then Nothing
+              else
+                if (Map.size (Map.filterWithKey (\k _ -> isWorkSpace firstModule k) (packages pkgLockV3)) > 0)
+                  then Just firstModule
+                  else Nothing
+
+        isWorkSpace :: Text -> PackagePath -> Bool
+        isWorkSpace name pkgPath = pkgPath == PackageLockV3WorkSpace name
+
+analyze :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs File -> m (Graphing Dependency)
+analyze file = context "Analyzing Npm Lockfile" $ do
+  packageLockJson <- context "Parsing v3 compatible package-lock.json" $ readContentsJson @PackageLockV3 file
+  context "Building dependency graph" $ pure $ buildGraph packageLockJson
+
+-- | Keeping repl'ble function in tactic for rapid debugging
+-- --
```

Comment:
Idk how OK with these - from my experience, npm is complexly subtle that we may have a few defects after initial support work. I would like to keep something like this for easy debugging, but OK with killing it as well - will defer to review.

Reply:
I'm fine with these, but let's add a `debugRepl` somewhere in the name, and annotate it with `{-# WARNING "some message" #-}` so that we can't accidentally use it.

We should make these easy to `grep` for, which is why I suggested `debugRepl`.  That way we can easily find them across the project for debugging.